snapcraft/internal/build_providers/_qemu/_qemu_driver.py

```diff
@@ -0,0 +1,175 @@
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import logging
+import shlex
+import shutil
+import socket
+import subprocess
+import sys
+import telnetlib
+import time
+from typing import List, Sequence
+
+import paramiko
+
+from snapcraft.internal.build_providers import errors
+
+
+logger = logging.getLogger(__name__)
+# Avoid getting paramiko logs which are overly verbose.
+logging.getLogger('paramiko').setLevel(logging.CRITICAL)
+
+
+def _popen(command: List) -> subprocess.Popen:
+    logger.debug('Running {}'.format(' '.join(command)))
+    return subprocess.Popen(command)
+
+
+def _get_qemu_command() -> str:
+    # TODO support more architectures.
+    return 'qemu-system-x86_64'
+
+
+class QemuDriver:
+    """A driver to interact with qemu virtual machines."""
+
+    provider_name = 'qemu'
+    _PROJECT_DEV = 'project_dev'
+    _PROJECT_MOUNT = 'project_mount'
+
+    def __init__(self, *, ssh_username: str, ssh_key_file: str) -> None:
+        """Initialize a QemuCommand instance.
+
+        :raises errors.ProviderCommandNotFound:
+            if the relevant qemu command is not found.
+        """
+        provider_cmd = _get_qemu_command()
+        if not shutil.which(provider_cmd):
+            raise errors.ProviderCommandNotFound(command=provider_cmd)
+        self.provider_cmd = provider_cmd
+        # TODO detect collisions and make dynamic
+        self._telnet_port = 64444
+        # TODO detect collisions and make dynamic
+        self._ssh_port = 5555
+        self._ssh_username = ssh_username
+        self._ssh_key_file = ssh_key_file
+        self._qemu_proc = None  # type: subprocess.Popen
+        self._ssh_handle = paramiko.SSHClient()
+        self._ssh_handle.set_missing_host_key_policy(paramiko.AutoAddPolicy())
+
+    def launch(self, *, hda: str, qcow2_drives: Sequence,
+               project_9p_dev: str, ram: str=None,
+               enable_kvm: bool=True) -> None:
+        # TODO check for latest snapshot to launch for it instead of a cold
+        #      boot.
+        # TODO replace -nographic with a spinner and redirect to a log file.
+        cmd = [
+            'sudo', self.provider_cmd,
+            '-m', ram, '-nographic',
+            '-monitor', 'telnet::{},server,nowait'.format(self._telnet_port),
+            '-hda', hda,
+            '-fsdev', 'local,id={},path={},security_model=none'.format(
+                self._PROJECT_DEV, project_9p_dev),
+            '-device', 'virtio-9p-pci,fsdev={},mount_tag={}'.format(
+                self._PROJECT_DEV, self._PROJECT_MOUNT),
+            '-device', 'e1000,netdev=net0',
+            '-netdev', 'user,id=net0,hostfwd=tcp::{}-:22'.format(
+                self._ssh_port)]
```

Comment:
Went doc-diving and found the `-nic` option, which I think covers both this and the above option in one go, while allowing you to drop some of the options (e.g. e1000 is the default):

>     -nic [tap|bridge|user|l2tpv3|vde|netmap|vhost-user|socket][,...][,mac=macaddr][,model=mn]
>
> This option is a shortcut for configuring both the on-board (default) guest NIC hardware and the host network backend in one go. The host backend options are the same as with the corresponding -netdev options below. The guest NIC model can be set with model=modelname. Use model=help to list the available device types. The hardware MAC address can be set with mac=macaddr.
>
> The following two example do exactly the same, to show how -nic can be used to shorten the command line length (note that the e1000 is the default on i386, so the model=e1000 parameter could even be omitted here, too):
>
>     qemu-system-i386 -netdev user,id=n1,ipv6=off -device e1000,netdev=n1,mac=52:54:98:76:54:32
>     qemu-system-i386 -nic user,ipv6=off,model=e1000,mac=52:54:98:76:54:32

Reply:
I thought that was deprecated, I'll check again. What I have I think is what multipass is using as well. I don't see where you do the host forwarding here though.

Reply:
I got confused with `-net` https://wiki.qemu.org/Documentation/Networking#The_legacy_-net_option
It seems `-nic` is not available on 18.04 yet though (sans the fact that it would miss the `hostfwd` part too.

Reply:
> The host backend options are the same as with the corresponding -netdev options below

That means to me that `hostfwd` would work here as well. But yeah, it not being available is a deal-killer.